onnx/onnx.proto

```diff
@@ -237,6 +245,172 @@ message TensorProto {
+
+  message KeyValuePairProto {
+    oneof key {
+      string s = 1;
+      int32 i32 = 2;
+      int64 i64 = 3;
+      uint64 ui64 = 4;
+    };
+    optional ValueProto value = 100;
+  }
+
+  message NameValuePairProto {
+    optional string key = 1;
+    optional ValueProto value = 100;
+  }
+
+  // Defines a record in its serialized format.
+  // A record is a sequence of one or more distinctly
+  // typed uniquely named values. 
+  message RecordProto {
+    repeated NameValuePairProto fields = 1;
+  }
+
+  // Defines a union value  in its serialized format.
+  // The name is the discriminator from the union type.
+  message UnionProto {
+    optional NamedValueProto choice = 1;
+  }
+  
+  // Defines a map in its serialized format.
+  // A record is a sequence of zero or more 
+  // key/value pairs all of which have unique
+  // keys
+  message MapProto {
+    repeated KeyValuePairProto key_value_pairs = 1;
+  }
+
+  // Alternate space-efficient encoding of map that MAY be used
+  // for maps whose value type is a scalar. 
+  message ScalarMapProto {
+    // keys.DataType must be an integral type or string
+    repeated TensorProto keys = 1;
+
+    // No restriction on data type, keys.length must equal values.length
+    repeated TensorProto values = 2;
+  }
+
+  // Defines a sequence in its serialized format.
+  // A sequence is a list of zero or more 
+  // tensors, maps, records, or subsequences.
+  message SequenceProto {
+    repeated ValueProto elems = 1;
+  }
+
+  oneof value {
+    // A dense tensor (or scalar).
+    optional TensorProto dense_tensor = 1;
+
+    // A sparse tensor.
+    optional SparseTensorProto sparse_tensor = 2;  
+
+    // A tuple.
+    optional RecordProto record = 3;
+
+    // A sequence.
+    optional SequenceProto seq = 4;
+
+    // A union.
+    optional UnionProto union = 5;
+
+    // A map.
+    optional MapProto map = 6;
+
+    // A map.
+    optional ScalarMapProto scalar_map = 7;
+  }
+}
+
+// Define the types.
+message TypeProto {
+  message TensorTypeProto {
+    // This field MUST NOT have the value of UNDEFINED
+    // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  message SparseTensorTypeProto {
+    // This field MUST NOT have the value of UNDEFINED
+    // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  // message T { ... }
+  message RecordTypeProto {
+  // The type and optional shape of each field 
+  // is described by a ValueInfoProto. The field
+  // names must be unique.     
+  // This field MUST be present for this version of the IR.
+    repeated ValueInfoProto field = 1;
+  };
+
+  // repeated T 
+  message SequenceTypeProto {
+  // The type and optional shape of each element of the sequence.
+  // This field MUST be present for this version of the IR.
+    optional TypeProto elem_type = 1;
+    optional TensorShapeProto elem_shape = 2;
```

Comment:
I am not sure how to interpret the elem_shape. I feel moving TensorShapeProto into TensorTypeProto will eliminate the need for this.

Reply:
(BTW, new to this thread, but I've been working with Yuan, Don, and other at Microsoft discussing this).

I agree with Rama, moving shape into the tensor type declaration would remove the need to have it appear wherever 'TypeProto' appears. It's really only applicable to tensors. In this case, it would appear that the shape is an attribute of the sequence, when in fact, it is an attribute of the element type.

Reply:
Notationally, moving TensorShapeProto inside of TensorTypeProto makes a ton of sense, as I can't imagine anywhere one would use TensorTypeProto and not have a TensorShapeProto on the side.

Yuan makes a valid point that shape inference can and will often happen distinctly from type inference. I don't think the notational change that you guys suggest impacts that.

Reply:
That's right, this is a syntactic change only, but it's one that makes it clear that shapes go with tensors, not with other types. Shape is a property of tensors, so it belongs in the type.

That shape inference and element type inference use separate logic and may flow independently should be impacted by this, as you say, notational change

Reply:
I wouldn't want to put both under the name of `TypeProto`, as it is more than type in the traditional sense. So if we want to put them together, how about naming it `TypeShapeProto`?

Reply:
I am fine with either name, but bundling them together into one unit makes sense to me

Reply:
That's fine with me. The name is less significant to me than having the two together as a unit.

Reply:
BTW, I opened issue #68 to track a decision on this.